doc/rst/technotes/errorHandling.rst

```diff
@@ -4,190 +4,581 @@
+      alwaysThrows(i);
+    } catch e {
+      writeln("Warning: caught a error ", e);
+    }
+  }
+
+2. ``try!`` instead of ``try``. This will halt the program if no matching
+``catch`` clause is found, instead of propagating.
+
+.. code-block:: chapel
+
+  proc haltsOnError(i: int): int {
+    try! {
+      canThrow(i);
+    } catch e: PrimerError {
+      writeln("caught a PrimerError");
+    }
+  }
+
+.. _technote-errorHandling-methods:
+
+Methods
+-------
+
+Errors can be thrown by methods, just as with any other function.
+An overriding method must throw if the overriden method throws,
+or not throw if the overridden method does not throw.
+
+.. code-block:: chapel
+
+  class ThrowingObject {
+    proc f() throws {
+      throw new Error();
     }
   }
 
-2. To the catch blocks of an enclosing ``try``.
+  class SubThrowingObject : ThrowingObject {
+    // must be marked throws even though it doesn't throw
+    proc f() throws {
+      writeln("this version doesn't throw");
+    }
+  }
+
+.. _technote-errorHandling-multilocale:
+
+Multilocale
+-----------
+
+Errors can be thrown within ``on`` statements. In that event, the error
+will be propagated out of the ``on`` statement.
 
 .. code-block:: chapel
 
-  try {
-    writeln("outer try");
+  proc handleFromOn() {
     try {
-      writeln("inner try");
-      throw new MyError();
-    } catch err: MyIntError {
-      writeln("MyError should not be caught here");
+      on Locales[0] {
+        canThrow(1);
+      }
+    } catch {
+      writeln("caught from Locale 0");
+    }
+  }
+
+.. _technote-errorHandling-parallel:
+
+Parallelism
+-----------
+
+``TaskErrors``
+++++++++++++++
+
+``TaskErrors`` helps coordinate errors among groups of tasks by collecting
+them for centralized handling. It can be iterated on and filtered for
+different kinds of errors.
+
+``begin``
++++++++++
+
+Errors can be thrown within a ``begin`` statement. In that event, the error
+will be propagated to the ``sync`` statement that waits for that task.
+
+.. code-block:: chapel
+
+  proc handleFromBegin() {
+    try! {
+      sync {
+        begin canThrow(0);
+        begin canThrow(1);
+      }
+    } catch e: TaskErrors {
+      writeln("caught from Locale 0");
+    }
+  }
+
+``coforall``, ``cobegin``
++++++++++++++++++++++++++
+
+Errors can be thrown from ``coforall`` and ``cobegin`` statements, handled
+as ``TaskErrors``.
+
+The implementation of ``TaskErrors`` prevents nested ``coforall`` statements
+from producing nested ``TaskErrors``. Instead, the nested errors will flatten
+into the outer loop's ``TaskErrors``.
+
+.. code-block:: chapel
+
+  proc handleFromCoforall() {
+    try! {
+      writeln("before coforall block");
+      coforall i in 1..2 {
+        coforall j in 1..2 {
+          throw new PrimerError();
+        }
+      }
+      writeln("after coforall block");
+    } catch errors: TaskErrors { // not nested
+      for e in errors {
+        writeln("Caught task error e ", e.message());
```

Comment:
Might be useful to put a comment here that e will always be of runtime type PrimerError (or whatever you rename it to)